airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialLoadStreamStateManager.java

```diff
@@ -0,0 +1,113 @@
+
+  // Map of pair to the primary key info (field name & data type) associated with it.
+  private final Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyInfo> pairToPrimaryKeyInfo;
+
+  private final Set<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair> streamsThatHaveCompletedSnapshot;
+  private static final Logger LOGGER = LoggerFactory.getLogger(MySqlInitialLoadStreamStateManager.class);
+
+  public MySqlInitialLoadStreamStateManager(final ConfiguredAirbyteCatalog catalog,
+                                            final InitialLoadStreams initialLoadStreams,
+                                            final Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyInfo> pairToPrimaryKeyInfo) {
+    this.pairToPrimaryKeyInfo = pairToPrimaryKeyInfo;
+    this.pairToPrimaryKeyLoadStatus = initPairToPrimaryKeyLoadStatusMap(initialLoadStreams.pairToInitialLoadStatus());
+    this.streamsThatHaveCompletedSnapshot = initStreamsCompletedSnapshot(initialLoadStreams, catalog);
+  }
+
+  @Override
+  public void updatePrimaryKeyLoadState(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair,
+                                        final PrimaryKeyLoadStatus pkLoadStatus) {
+    pairToPrimaryKeyLoadStatus.put(pair, pkLoadStatus);
+  }
+
+  public AirbyteStateMessage createFinalStateMessage(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair,
+                                                     final JsonNode streamStateForIncrementalRun) {
+    streamsThatHaveCompletedSnapshot.add(pair);
+
+    return new AirbyteStateMessage()
+        .withType(AirbyteStateType.STREAM)
+        .withData(Jsons.jsonNode(getFinalState(pair)))
+        .withStream(getAirbyteStreamState(pair, (streamStateForIncrementalRun)));
+  }
+
+  @Override
+  public PrimaryKeyInfo getPrimaryKeyInfo(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair) {
+    return pairToPrimaryKeyInfo.get(pair);
+  }
+
+  @Override
+  public PrimaryKeyLoadStatus getPrimaryKeyLoadStatus(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair) {
+    return pairToPrimaryKeyLoadStatus.get(pair);
+  }
+
+  @Override
+  public AirbyteStateMessage createIntermediateStateMessage(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair,
+                                                            final PrimaryKeyLoadStatus pkLoadStatus) {
+    return new AirbyteStateMessage()
+        .withType(AirbyteStateType.STREAM)
+        .withStream(getAirbyteStreamState(pair, Jsons.jsonNode(pkLoadStatus)));
+  }
+
+
+
+  private AirbyteStreamState getAirbyteStreamState(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair, final JsonNode stateData) {
+    LOGGER.info("STATE DATA FOR {}: {}", pair.getNamespace().concat("_").concat(pair.getName()), stateData);
+    assert Objects.nonNull(pair);
```

Comment:
`pair` will never be null out here since we are using it in the logger statement above, if its null it will throw a NPE before getting here